Wallet: estimate the size of signed inputs using descriptors

darosior opened the pull request:
The wallet currently estimates the size of a signed input by doing a dry run of the signing logic. This is unnecessary since all outputs we can sign for can be represented by a descriptor, and we can derive the size of a satisfaction ("signature") directly from the descriptor itself.
In addition, the current approach does not generalize well: dry runs of the signing logic are only possible for the most basic scripts. See for instance the discussion in #24149 around that.

This introduces a method to get the maximum size of a satisfaction from a descriptor, and makes the wallet use that instead of the dry-run.

darosior committed:
descriptor: introduce a method to get the satisfaction size

In the wallet code, we are currently estimating the size of a signed
input by doing a dry run of the signing logic. This is unnecessary as
all outputs we are able to sign for can be represented by a descriptor,
and we can derive the size of a satisfaction ("signature") from the
descriptor itself directly.
In addition, this approach does not scale: getting the size of a
satisfaction through a dry run of the signing logic is only possible for
the most basic scripts.

This commit introduces the computation of the size of satisfaction per
descriptor. It's a bit intricate for 2 main reasons:
- We want to conserve the behaviour of the current dry-run logic used by
  the wallet that sometimes assumes ECDSA signatures will be low-r,
  sometimes not (when we don't create them).
- We need to account for the witness discount. A single descriptor may
  sometimes benefit of it, sometimes not (for instance `pk()` if used as
  top-level versus if used inside `wsh()`).

darosior committed:
wallet: use descriptor satisfaction size to estimate inputs size

Instead of using the dummysigner to compute a placeholder satisfaction,
infer a descriptor on the scriptPubKey of the coin being spent and use
the estimation of the satisfaction size given by the descriptor
directly.

Note this (almost, see next paragraph) exactly conserves the previous
behaviour. For instance CalculateMaximumSignedInputSize was previously
assuming the input to be spent in a transaction that spends at least one
Segwit coin, since it was always accounting for the serialization of the
number of witness elements.

In this commit we use a placeholder for the size of the serialization of
the witness stack size (1 byte). Since the logic in this commit is
already tricky enough to review, and that it is only a very tiny
approximation not observable through the existing tests, it is addressed
in the next commit.

darosior committed:
wallet: accurately account for the size of the witness stack

When estimating the maximum size of an input, we were assuming the
number of elements on the witness stack could be encode in a single
byte. This is a valid approximation for all the descriptors we support
(including P2WSH Miniscript ones), but may not hold anymore once we
support Miniscript within Taproot descriptors (since the max standard
witness stack size of 100 gets lifted).

It's a low-hanging fruit to account for it correctly, so just do it now.

darosior commented:
Pushed a commit to accurately compute the serialized size of the witness stack size, cleanup the commits a bit, and add the newly introduced methods in the descriptors to the `descriptor_parse` fuzz target. This is now ready for review.

darosior commented on src/script/descriptor.cpp:
```diff
@@ -931,6 +1058,20 @@ class TRDescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> MaxSatisfactionWeight(bool) const override {
+        // FIXME: We assume keypath spend, which can lead to very large underestimations.
+        return 1 + 64;
```
Since we want the maximum size, we should probably account for the witness byte and make it `65` instead.

darosior replied:
Done.

darosior committed:
miniscript: make GetStackSize independent of P2WSH context

It was taking into account the P2WSH script push in the number of stack
elements.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -740,6 +755,17 @@ class PKDescriptor final : public DescriptorImpl
     PKDescriptor(std::unique_ptr<PubkeyProvider> prov, bool xonly = false) : DescriptorImpl(Vector(std::move(prov)), "pk"), m_xonly(xonly) {}
     bool IsSingleType() const final { return true; }
+
+    std::optional<int64_t> ScriptSize() const override { return 1 + (m_xonly ? 33 : 32) + 1; }
```
In c8b0aeb534edce08366176275d5577c1e309eb6f "descriptor: introduce a method to get the satisfaction size"

This is backwards, should be 32 for `m_xonly`.

```suggestion
    std::optional<int64_t> ScriptSize() const override { return 1 + (m_xonly ? 32 : 33) + 1; }
```

darosior replied:
Woops. Thanks.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -740,6 +755,17 @@ class PKDescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> MaxSatSize(bool use_max_sig) const override {
+        const auto ecdsa_sig_size = use_max_sig ? 72 : 71;
+        return 1 + (m_xonly ? ecdsa_sig_size : 64);
```
In c8b0aeb534edce08366176275d5577c1e309eb6f "descriptor: introduce a method to get the satisfaction size"

This is backwards, should be 64 for `m_xonly`.

```suggestion
        return 1 + (m_xonly ? 64 : ecdsa_sig_size);
```

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -756,6 +782,17 @@ class PKHDescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> MaxSatSize(bool use_max_sig) const override {
+        const auto sig_size = use_max_sig ? 72 : 71;
+        return 1 + sig_size + 1 + 33;
```
In c8b0aeb534edce08366176275d5577c1e309eb6f "descriptor: introduce a method to get the satisfaction size"

We could have an uncompressed pubkey, so we should get the size from the `PubkeyProvider`.

darosior replied:
Ah good catch! Done, thanks.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -816,6 +864,20 @@ class MultisigDescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> ScriptSize() const override {
+        const auto n_keys = m_pubkey_args.size();
+        return 1 + BuildScript(n_keys).size() + BuildScript(m_threshold).size() + 34 * n_keys;
```
In c8b0aeb534edce08366176275d5577c1e309eb6f "descriptor: introduce a method to get the satisfaction size"

Same uncompressed pubkey issue as above.

achow101 replied:
This wasn't changed?

darosior replied:
Oops. Done now.

achow101 reviewed (commented):
Concept ACK

darosior commented:
Thanks for the review. Addressed your comments and rebased on master.

darosior commented on src/wallet/spend.cpp:
```diff
@@ -46,17 +90,65 @@ int CalculateMaximumSignedInputSize(const CTxOut& txout, const CWallet* wallet,
+    }
+    if (coin_control) {
+        return InferDescriptor(script_pubkey, coin_control->m_external_provider);
+    }
```
Hmm good point. Unfortunately this is a bit clumsy for the `LegacyScriptPubKeyMan`: since it's not a `FlatSigningProvider` it involves copying all its data in order to merge both providers. I've pushed the change in a separate commit to be squashed if it's not too unreasonable. What do you think?

fanquake replied:
> What do you think?

@achow101 followup thoughts here?

darosior replied:
Pushed a much cleaner version of this: introducing a `MultiSigningProvider` to wrap around both the wallet and external providers. Still as a commit on top, let me know what you think @achow101.

darosior committed:
miniscript: introduce a helper to get the maximum witness size

Similarly to how we compute the maximum stack size.

Also note how it would be quite expensive to recompute it recursively
by accounting for different ECDSA signature sizes. So we just assume
high-R everywhere. It's only a trivial difference anyways.

darosior commented:
Rebased, and added Miniscript support since signing support was merged.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -843,6 +907,15 @@ class MultiADescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> MaxSatSize(bool use_max_sig) const override {
+        return (1 + 65) * m_threshold + (m_pubkey_args.size() - m_threshold);
+    }
```
In 9e8df2de1750b823492c18a924fefbd7d474819e "descriptor: introduce a method to get the satisfaction size"

Is `MultiADescriptor` supposed to be missing a `MaxSatisfactionWeight`?

darosior replied:
Yes. It would never be used as `MaxSatisfactionWeight` is only called on top-level descriptors.

achow101 reviewed (commented):
The `MultiSigningProvider` seems like a fine approach.

darosior committed:
script/signingprovider: introduce a MultiSigningProvider

It is sometimes useful to interface with multiple signing providers at
once. For instance when inferring a descriptor with solving information
being provided from multiple sources (see next commit).

Instead of inneficiently copying the information from one provider into
the other, introduce a new signing provider that takes a list of
pointers to existing providers.

darosior commented:
Rebased, added more test coverage (we must be able to estimate the satisfaction size for any solvable top-level descriptor), properly integrated the `MultiSigningProvider` and removed the `MaxSatisfactionWeight` for `MiniscriptDescriptor` (which would never be called as they are never top-level).

achow101 commented on src/script/miniscript.h:
```diff
@@ -846,6 +857,56 @@ struct Node {
+            }
+            case Fragment::OR_C: return {subs[0]->ws.sat | (subs[0]->ws.dsat + subs[1]->ws.sat), {}};
+            case Fragment::OR_D: return {subs[0]->ws.sat | (subs[0]->ws.dsat + subs[1]->ws.sat), subs[0]->ws.dsat + subs[1]->ws.dsat};
+            case Fragment::OR_I: return {(subs[0]->ws.sat + 1 + 1) | (subs[1]->ws.sat + 1), (subs[0]->ws.dsat + 1 + 1) | (subs[1]->ws.dsat + 1)};
```
In 3e3521bf956c4387abe4b6ce12ce55224a9e97ba "miniscript: introduce a helper to get the maximum witness size"

I don't quite follow why `subs[0]` needs an additional `+ 1` that `subs[1]` doesn't have. AFAICT, `subs[0]` just has a `1` in the witness, while `subs[1]` has a `0`, and both should be the same size.

darosior replied:
Because serializing a `1` as a witness stack element needs to specify a 1-byte long element and then the value of this byte (serialized `0x0101`) whereas encoding a `0` is simply the empty witness (serialized `0x00`).

achow101 replied:
Oh right, duh.

darosior commented:
Rebased.

darosior commented on src/script/descriptor.cpp:
```diff
@@ -1040,6 +1156,14 @@ class MiniscriptDescriptor final : public DescriptorImpl
+    std::optional<int64_t> ScriptSize() const override { return m_node->ScriptSize(); }
+
+    std::optional<int64_t> MaxSatSize(bool) const override {
+        // We only ever parse satisfiable Miniscripts. Also note for Miniscript we always assume
+        // high-R ECDSA signatures.
+        return CHECK_NONFATAL(m_node->GetWitnessSize());
```
This is not true if the Miniscript descriptor was parsed from Script. I'll update this.

darosior replied:
Done.

achow101 commented on src/script/miniscript.h:
```diff
@@ -1152,12 +1152,12 @@ struct Node {
     std::optional<uint32_t> GetStackSize() const {
         if (!ss.sat.valid) return {};
-        return ss.sat.value + 1;
+        return ss.sat.value;
```
In ecdee0c3f802c6308c358a5ee501d2e031c8dcf2 "miniscript: make GetStackSize independent of P2WSH context"

The comment above this function is outdated.

darosior replied:
Updated the comment.

darosior commented:
Rebased after #27888.

darosior commented:
Rebased after #28244.

darosior commented:
Hey i keep rebasing this, i felt it was quite close during the last reviews from @achow101. Would be nice to have others look at this though.

@Sjors and @furszy would you mind having a look here? I think you'd be the best candidates for the job.

achow101 commented on src/script/miniscript.h:
```diff
@@ -1148,8 +1209,8 @@ struct Node {
-    /** Return the maximum number of stack elements needed to satisfy this script non-malleably, including
-     * the script push. */
+    /** Return the maximum number of stack elements needed to satisfy this script non-malleably.
+     * This does not account for the P2WSH script push. */
```
In a7db3f0503626c3671482720b22fedef7cd9e23c "miniscript: introduce a helper to get the maximum witness size"

nit: These comments should be changed in the previous commit.

darosior replied:
Done.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -50,15 +96,67 @@ int CalculateMaximumSignedInputSize(const CTxOut& txout, const CWallet* wallet,
+    // Otherwise, use the maximum satisfaction size provided by the descriptor.
+    std::unique_ptr<Descriptor> desc{GetDescriptor(wallet, coin_control, txo.scriptPubKey)};
+    if (desc) {
+        if (const auto weight = MaxInputWeight(*desc, {txin}, coin_control, tx_is_segwit, can_grind_r)) return *weight;
```
In 310a7382454be01262b3fb2b089372a71a8b4482 "wallet: use descriptor satisfaction size to estimate inputs size"

nit: Could just return the result of `MaxInputWeight` directly.

darosior replied:
Done.

achow101 reviewed (commented):
ACK 932556c87f939ffcdf1d591075d4c54bbd9bcaea

furszy reviewed (commented):
Concept ACK

Added to my queue. Will start reviewing next week.

darosior commented on src/test/descriptor_tests.cpp:
```diff
@@ -150,6 +150,13 @@ void DoCheck(std::string prv, std::string pub, const std::string& norm_pub, int
     parse_pub = Parse(pub, keys_pub, error);
     BOOST_CHECK_MESSAGE(parse_pub, error);
 
+    // We must be able to estimate the max satisfaction size for any solvable descriptor top descriptor (but combo).
```
Good call, done.

darosior commented on src/test/fuzz/descriptor_parse.cpp:
```diff
@@ -132,6 +132,13 @@ static void TestDescriptor(const Descriptor& desc, FlatSigningProvider& sig_prov
     if (!out_scripts.empty()) {
         assert(InferDescriptor(out_scripts.back(), sig_provider));
     }
+
```
Done.

darosior commented on src/wallet/spend.cpp:
```diff
@@ -50,15 +96,67 @@ int CalculateMaximumSignedInputSize(const CTxOut& txout, const CWallet* wallet,
-    return TxSize{vsize, weight};
+    // nVersion + nLockTime + input count + output count
+    int64_t weight = (4 + 4 + GetSizeOfCompactSize(tx.vin.size()) + GetSizeOfCompactSize(tx.vout.size())) * WITNESS_SCALE_FACTOR;
+    // Whether any input spend a witness program. Necessary to run before the next loop over the
```
Done.

darosior commented on src/wallet/spend.cpp:
```diff
@@ -34,14 +34,60 @@ using interfaces::FoundBlock;
+static std::optional<int64_t> MaxInputWeight(const Descriptor& desc, const std::optional<CTxIn>& txin,
+                                             const CCoinControl* coin_control, const bool tx_is_segwit,
+                                             const bool can_grind_r) {
+    if (const auto sat_weight = desc.MaxSatisfactionWeight(can_grind_r && UseMaxSig(txin, coin_control))) {
```
Good catch, fixed. I was a bit concerned that this wouldn't make any test fail, but it's simply because those use wallets without external signer (so `can_grind_r` is always `true`). Actually getting the condition backward:
```patch
diff --git a/src/wallet/spend.cpp b/src/wallet/spend.cpp
index 750b6c100b..45f225b18f 100644
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -58,7 +58,7 @@ static bool UseMaxSig(const std::optional<CTxIn>& txin, const CCoinControl* coin
 static std::optional<int64_t> MaxInputWeight(const Descriptor& desc, const std::optional<CTxIn>& txin,
                                              const CCoinControl* coin_control, const bool tx_is_segwit,
                                              const bool can_grind_r) {
-    if (const auto sat_weight = desc.MaxSatisfactionWeight(!can_grind_r || UseMaxSig(txin, coin_control))) {
+    if (const auto sat_weight = desc.MaxSatisfactionWeight(can_grind_r && !UseMaxSig(txin, coin_control))) {
         if (const auto elems_count = desc.MaxSatisfactionElems()) {
             const bool is_segwit = IsSegwit(desc);
             // Account for the size of the scriptsig and the number of elements on the witness stack. Note
```
 would make the tests for this feature fail.

darosior commented on src/script/descriptor.cpp:
```diff
@@ -905,6 +1016,26 @@ class WSHDescriptor final : public DescriptorImpl
+    std::optional<int64_t> MaxSatSize(bool use_max_sig) const override {
+        if (const auto sat_size = m_subdescriptor_args[0]->MaxSatSize(use_max_sig)) {
+            if (const auto subscript_size = m_subdescriptor_args[0]->ScriptSize()) {
+                return 1 + *subscript_size + *sat_size;
```
Thanks, done. This one slipped through my adapting this to consider Miniscript descriptors.

darosior commented on src/script/descriptor.cpp:
```diff
@@ -958,6 +1067,13 @@ class TRDescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> MaxSatisfactionWeight(bool) const override {
+        // FIXME: We assume keypath spend, which can lead to very large underestimations.
+        return 1 + 65;
```
Since this is estimating the maximum, i figured we shouldn't assume SIGHASH_DEFAULT. Do you think we should?

darosior commented on src/wallet/spend.cpp:
```diff
@@ -34,14 +34,60 @@ using interfaces::FoundBlock;
-        return -1;
+    if (!provider) return -1;
+
+    if (const auto desc = InferDescriptor(txout.scriptPubKey, *provider)) {
```
This wouldn't miss any case that were handled previously as far as i can tell. `DummySignInput` would previously return `false` on a P2PKH for which we don't know the public key. However there may be ways to improve upon the current situation, for instance we could set `MaxInputWeight` for an `addr()`/`raw()` descriptor that represents a P2(W)PKH, since we know it? But i'd prefer to first get this switch to using descriptors done with no regression and leave improvements for a follow-up, if reviewers don't mind.

darosior commented on src/script/descriptor.cpp:
```diff
@@ -1059,6 +1182,13 @@ class RawTRDescriptor final : public DescriptorImpl
+
+    std::optional<int64_t> MaxSatisfactionWeight(bool) const override {
+        // We can't know whether there is a script path, so assume key path spend.
+        return 1 + 65;
```
Resolving for the same reason as https://github.com/bitcoin/bitcoin/pull/26567#discussion_r1307483129.

achow101 commented:
re-ACK 10546a569c6c96a5ec1b9708abf9ff5c8644f669
